### DISCUSSION TITLE
Server Constantly Shutting Down

### DISCUSSION THREAD

**Discussion by AndyK28** [getgauge/gauge#2632] [General]
I constantly getting this error whenever working on my specs:
Client Gauge: connection to server is erroring. Shutting down server.

This is really frustrating as I have to restart vscode time and time again to get gauge running again, this issue brings down gauge intellisense and sometimes my specs come out as red even though there is no error.

I am using the latest version of gauge:
```bash
Gauge version: 1.6.9
Commit Hash: aff43ef

Plugins
-------
dotnet (0.7.2)
html-report (4.3.1)
screenshot (0.3.0)
spectacle (0.2.0)
```

my vscode gauge extension version is v0.1.6

**Comment by chadlwilson** [comment]
What's in the gauge server logs?

**Reply by AndyK28** [reply]
I will have to wait for it to crash again then I will share with you

**Reply by chadlwilson** [reply]
I don't think the logs get deleted each time, so not sure you have to wait for that (unless you deleted them).

**Reply by AndyK28** [reply]
Ever since I posted this here I haven't had the issue for some weird reason and I didnt see your response on time to get the logs and tracking them now would be time consuming for me. But thank you for your prompt response. 👍🏾

**Comment by AndyK28** [comment]
Below is the attached logs for the server crash

![Screenshot 2024-10-29 114232](https://github.com/user-attachments/assets/025fcadc-e285-421d-95a2-e35a84627ec9)
![Screenshot 2024-10-29 115222](https://github.com/user-attachments/assets/8a9f1aa0-3780-49a0-ad8f-b19ac1cb8aad)

**Reply by chadlwilson** [reply]
It looks like there is some kind a bug during trying to offer step text completion options to the IDE while you are working on a `.spec`.

https://github.com/getgauge/gauge/blob/35a13be9335c919f6eefdcb15ba6688846044de2/api/lang/completionStep.go#L61-L72

On Line 69, probably an error being returned during `parser.ExtractStepValueAndParams(stepText, false)`

I'm not sure how to replicate it (this code isn't really my core expertise) so not sure what I can do here. Would probably need an example `.spec` file with some content that can reproduce the problem.

The code doesn't seem to handle or ignore errors from that parse, so guess if there **is** an error returned while parsing it will crash, rather than ignore the error and not give a completion.

**Reply by AndyK28** [reply]
I would provide something for you to work with but the issue is I too don't know what causes the issue or when exactly the issue occurs. For now it's all random and I can't tell at what point does the server crash.

**Reply by sriv** [reply]
I raised a PR here - https://github.com/getgauge/gauge/pull/2638

this should log an error and proceed without crashing then step parsing fails.

**Reply by AndyK28** [reply]
Thanks, will wait for the PR to be approved and we'll see with the release if the issue persists.

**Comment by AndyK28** [comment]
I figured one of the reasons for the server crash. After careful observation I noticed that after writing or modifying a step and you move on to a new one then leave the asterisk blank without step declaration the server crashes. I reproduced this 5 times and I am getting the same results. Give it approx 10 - 15 seconds then it will crash. Not sure if this is the only cause though

![image](https://github.com/user-attachments/assets/ff555774-887d-43ae-bd62-e37740f89f5b)

**Reply by sriv** [reply]
are there any more details in the log? perhaps an error message or stack trace?

**Reply by chadlwilson** [reply]
Can you upgrade to Gauge 1.6.10 and try that?

It may or may not crash, but in either case if you can help us see what it now logs it'd be useful.

**Reply by AndyK28** [reply]
Okay give me a moment to update

**Reply by AndyK28** [reply]
With the update I am not getting a server crash as yet and I am noticing that 1.6.9 didn't recognise steps as steps but just a text picked up by my intellisense if not gauge itself.

**Reply by AndyK28** [reply]
I am signing out in 10 mins. Will sure let you know should I encounter a similar issue as the week progresses. 👍🏾

**Reply by chadlwilson** [reply]
If you are triggering the same problem as earlier (the only one that would have any change in behaviour)  you should see "Unable to extract StepValueAndParams* in your logs. That's the log entry that we'd probably need the message from to address the root problem.

**Reply by AndyK28** [reply]
Here are the logs:

```
29-10-2024 16:56:31.655 [Gauge] [DEBUG] request #67: textDocument/codeLens: failed to parse d:\\Specs\Dashboard\05_AssetStatus.spec
d:\\Specs\Dashboard\05_AssetStatus.spec:27 Step should not be blank
29-10-2024 16:56:31.655 [Gauge] [DEBUG] jsonrpc2: <-- notif: window/logMessage: {"type":4,"message":"request #67: textDocument/codeLens: failed to parse d:\\Specs\\Dashboard\\05_AssetStatus.spec\nd:\\Specs\\Dashboard\\05_AssetStatus.spec:27 Step should not be blank\n"}
29-10-2024 16:56:31.656 [Gauge] [DEBUG] jsonrpc2: <-- error #67: textDocument/codeLens: {"code":0,"message":"failed to parse d:\\Specs\\Dashboard\\05_AssetStatus.spec\nd:\\Specs\\Dashboard\\05_AssetStatus.spec:27 Step should not be blank\n"}
29-10-2024 16:56:35.929 [Gauge] [DEBUG] jsonrpc2: --> request #68: textDocument/codeAction: {"context":{"diagnostics":[],"triggerKind":2},"range":{"end":{"character":31,"line":25},"start":{"character":31,"line":25}},"textDocument":{"uri":"file:///d%!A(MISSING)/Specs/Dashboard/05_AssetStatus.spec"}}
29-10-2024 16:56:35.929 [Gauge] [DEBUG] jsonrpc2: <-- result #68: textDocument/codeAction: null
29-10-2024 16:56:36.668 [Gauge] [DEBUG] jsonrpc2: --> request #69: textDocument/codeAction: {"context":{"diagnostics":[{"message":"Step should not be blank","range":{"end":{"character":1,"line":26},"start":{"character":0,"line":26}},"severity":1}],"triggerKind":2},"range":{"end":{"character":1,"line":26},"start":{"character":1,"line":26}},"textDocument":{"uri":"file:///d%!A(MISSING)/Specs/Dashboard/05_AssetStatus.spec"}}
29-10-2024 16:56:36.668 [Gauge] [DEBUG] jsonrpc2: <-- result #69: textDocument/codeAction: null
29-10-2024 16:56:37.800 [Gauge] [DEBUG] jsonrpc2: --> request #70: textDocument/codeAction: {"context":{"diagnostics":[{"message":"Step should not be blank","range":{"end":{"character":1,"line":26},"start":{"character":0,"line":26}},"severity":1}],"triggerKind":2},"range":{"end":{"character":1,"line":26},"start":{"character":31,"line":25}},"textDocument":{"uri":"file:///d%!A(MISSING)/Specs/Dashboard/05_AssetStatus.spec"}}
29-10-2024 16:56:37.800 [Gauge] [DEBUG] jsonrpc2: <-- result #70: textDocument/codeAction: null
29-10-2024 16:56:38.378 [Gauge] [DEBUG] jsonrpc2: --> notif: textDocument/didChange: {"contentChanges":[{"text":"# Dashboard AssetStatus Feature\r\nTags: 02_DashboardFeatures, assetStatus\r\n\r\n## Last Position Asset Status Check\r\nTags: lastPositionCount\r\n\r\n* Navigate To Dashboard Interface.\r\n* Click on tree account \"AccountName\".\r\n* Verify that the last position count matches database count.\r\n\r\n## Refresh Last Position Status Check\r\nTags: refreshLastPosition\r\n\r\n* Navigate To Dashboard Interface.\r\n* Expand all accounts in the asset tree.\r\n* Click on tree asset \"KG09YFGP\".\r\n* Verify that the selected asset appears as the only asset in the Last Positions Table.\r\n* Refresh last position.\r\n* Verify that the last position count matches the last position table row count after refresh.\r\n\r\n## Select Assets To Put In Focus\r\nTags: inFocus\r\n\r\n* Navigate To Dashboard Interface.\r\n* Select assets \"3\" to set in focus.\r\n* Focus on the selected assets."}],"textDocument":{"uri":"file:///d%!A(MISSING)/Specs/Dashboard/05_AssetStatus.spec","version":12}}
29-10-2024 16:56:38.379 [Gauge] [DEBUG] jsonrpc2: --> request #71: textDocument/completion: {"context":{"triggerKind":1},"position":{"character":31,"line":25},"textDocument":{"uri":"file:///d%!A(MISSING)/Specs/Dashboard/05_AssetStatus.spec"}}
29-10-2024 16:56:38.381 [Gauge] [ERROR] Unable to extract StepValueAndParams for step '@"\(\d+\)', error : String not terminated
29-10-2024 16:56:38.381 [Gauge] [DEBUG] jsonrpc2: <-- notif: window/logMessage: {"type":1,"message":"Unable to extract StepValueAndParams for step '@\"\\(\\d+\\)', error : String not terminated"}
29-10-2024 16:56:38.381 [Gauge] [ERROR] Unable to extract StepValueAndParams for step '@"\((\d+)\)', error : String not terminated
29-10-2024 16:56:38.381 [Gauge] [DEBUG] jsonrpc2: <-- notif: window/logMessage: {"type":1,"message":"Unable to extract StepValueAndParams for step '@\"\\((\\d+)\\)', error : String not terminated"}
29-10-2024 16:56:38.381 [Gauge] [ERROR] Unable to extract StepValueAndParams for step '@"\d+', error : String not terminated
29-10-2024 16:56:38.381 [Gauge] [DEBUG] jsonrpc2: <-- notif: window/logMessage: {"type":1,"message":"Unable to extract StepValueAndParams for step '@\"\\d+', error : String not terminated"}
29-10-2024 16:56:38.381 [Gauge] [ERROR] Unable to extract StepValueAndParams for step '@"\s*\(.*\)$', error : String not terminated
29-10-2024 16:56:38.381 [Gauge] [DEBUG] jsonrpc2: <-- notif: window/logMessage: {"type":1,"message":"Unable to extract StepValueAndParams for step '@\"\\s*\\(.*\\)$', error : String not terminated"}
```

**Reply by AndyK28** [reply]
> With the update I am not getting a server crash as yet and I am noticing that 1.6.9 didn't recognise steps as steps but just a text picked up by my intellisense if not gauge itself.

Here's what I was talking about when it came to steps:

V1.6.9
![Screenshot 2024-10-30 083930](https://github.com/user-attachments/assets/299db7ed-cfeb-4de2-85af-c7871c0e9294)


V1.6.10
![Screenshot 2024-10-30 084257](https://github.com/user-attachments/assets/ff27840e-a503-4bf2-97d3-14b71341c26a)

**Reply by chadlwilson** [reply]
From a quick look this seems like the type of error which can be expected when youve got work in progress on a spec/step and Gauge should handle it more cleanly (as 1.6.10 now does) but don't quite understand why the crash wasn't more widespread earlier. 🤔

**Reply by AndyK28** [reply]
Remains a mystery I guess...

**Comment by AndyK28** [comment]
I think the logs are fine now


![Screenshot 2024-10-29 165246](https://github.com/user-attachments/assets/92cc3aba-629f-480d-864b-f00b5c3a7926)
